modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/MaskOverviewProvider.java

```diff
@@ -0,0 +1,692 @@
+    }
+
+    public ImageReaderSpi getMaskReaderSpi() {
+        return maskReaderSpi;
+    }
+
+    public ImageInputStreamSpi getMaskOvrStreamSpi() {
+        return maskOvrStreamSpi;
+    }
+
+    public ImageReaderSpi getMaskOvrReaderSpi() {
+        return maskOvrReaderSpi;
+    }
+
+    public DatasetLayout getLayout() {
+        return layout;
+    }
+
+    public int getNumOverviews() {
+        return numOverviews;
+    }
+
+    public int getNumInternalOverviews() {
+        return numInternalOverviews;
+    }
+
+    public int getNumExternalOverviews() {
+        return numExternalOverviews;
+    }
+
+    public int getNumInternalMasks() {
+        return numInternalMasks;
+    }
+
+    public int getNumExternalMasks() {
+        return numExternalMasks;
+    }
+
+    public int getNumExternalMasksOverviews() {
+        return numExternalMasksOverviews;
+    }
+
+    public URL getFileURL() {
+        return fileURL;
+    }
+
+    public URL getOvrURL() {
+        return ovrURL;
+    }
+
+    public URL getMaskURL() {
+        return maskURL;
+    }
+
+    public URL getMaskOvrURL() {
+        return maskOvrURL;
+    }
+
+    /**
+     * Returns an {@link ImageInputStreamSpi} instance for the input {@link URL}
+     */
+    public static ImageInputStreamSpi getInputStreamSPIFromURL(URL fileURL) throws IOException {
+
+        ImageInputStreamSpi streamSPI = ImageIOExt.getImageInputStreamSPI(fileURL, true);
+        if (streamSPI == null) {
+            final File file = DataUtilities.urlToFile(fileURL);
+            if (file != null) {
+                if (LOGGER.isLoggable(Level.WARNING)) {
+                    LOGGER.log(Level.WARNING, file.getCanonicalPath());
+                }
+            }
+            throw new IllegalArgumentException(
+                    "Unable to get an input stream for the provided source " + fileURL.toString());
+        }
+        return streamSPI;
+    }
+
+    /**
+     * Returns an {@link ImageReaderSpi} instance for the input {@link ImageInputStream} and the optional suggested spi.
+     */
+    public static ImageReaderSpi getReaderSpiFromStream(ImageReaderSpi suggestedSPI,
+            ImageInputStream inStream) throws IOException {
+        ImageReaderSpi readerSPI = null;
+        // get a reader and try to use the suggested SPI first
+        inStream.mark();
+        if (suggestedSPI != null && suggestedSPI.canDecodeInput(inStream)) {
+            readerSPI = suggestedSPI;
+            inStream.reset();
+        } else {
+            inStream.mark();
+            ImageReader reader = ImageIOExt.getImageioReader(inStream);
+            if (reader != null)
+                readerSPI = reader.getOriginatingProvider();
+            inStream.reset();
+        }
+        return readerSPI;
+    }
+
+    /**
+     * Returns a {@link ROI} object based on the input {@link RenderedImage} representing ROI
+     */
+    public static ROI postProcessROI(RenderedImage roiRaster, Rectangle bounds) {
```

Comment:
This looks more like a scaleROI method to me (postProcess could be ... anything, this one does a scaling)